components/layout/traversal.rs

```diff
@@ -167,13 +169,31 @@ pub(crate) fn compute_damage_and_repair_style_inner(
     if element_damage != RestyleDamage::reconstruct() &&
         damage_for_parent.contains(RestyleDamage::RELAYOUT)
     {
-        node.with_each_layout_box_base_including_pseudos(
-            if (original_element_damage | damage_from_children).contains(RestyleDamage::RELAYOUT) {
-                LayoutBoxBase::clear_fragments_and_layout_cache
-            } else {
-                LayoutBoxBase::clear_fragments
-            },
-        );
+        let outer_inline_content_sizes_depend_on_content = Cell::new(false);
+        node.with_each_layout_box_base_including_pseudos(|base| {
+            base.clear_fragments();
+            if original_element_damage.contains(RestyleDamage::RELAYOUT) {
+                *base.cached_layout_result.borrow_mut() = None;
+                *base.cached_inline_content_size.borrow_mut() = None;
+            } else if damage_from_children.contains(RestyleDamage::RELAYOUT) {
+                *base.cached_layout_result.borrow_mut() = None;
+                if !damage_from_children.contains(LayoutDamage::recompute_inline_content_sizes()) {
+                    return;
+                }
+                *base.cached_inline_content_size.borrow_mut() = None;
+            }
+            if !base.base_fragment_info.is_anonymous() {
+                // TODO: Use `Cell::update()` once it becomes stable.
+                outer_inline_content_sizes_depend_on_content.set(
+                    outer_inline_content_sizes_depend_on_content.get() ||
+                        base.outer_inline_content_sizes_depend_on_content
+                            .load(Ordering::Relaxed),
+                );
+            }
+        });
+        if outer_inline_content_sizes_depend_on_content.get() {
+            damage_for_parent.insert(LayoutDamage::recompute_inline_content_sizes())
+        }
```

Comment:
```suggestion
        // Only clear the cached inline content sizes on ancestors, when this node's inline content
        // sizes actually depend on content, otherwise the ancestors don't on the descendant's
        // content either.
        if outer_inline_content_sizes_depend_on_content.get() {
            damage_for_parent.insert(LayoutDamage::recompute_inline_content_sizes())
        }
```